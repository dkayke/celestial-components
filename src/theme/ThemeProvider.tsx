import React, { useContext } from 'react'
import { ThemeProvider as SCThemeProvider } from 'styled-components'
import celestialTheme from './'
import { ThemeProviderProps } from './theme.d'
import { ThemeContext, ThemeContextProvider } from './ThemeContext'

type DeepPartial<T> = {
  [K in keyof T]?: T[K] extends object ? DeepPartial<T[K]> : T[K]
}

const merge = (theme: Theme, newTheme?: DeepPartial<Theme>): Theme => {
  return {
    color: {
      primary: { ...theme.color.primary, ...newTheme?.color?.primary }
    },
    font: { ...theme.font, ...newTheme?.font }
  }
}

const ThemeProvider: React.FC<ThemeProviderProps> = ({ children, theme }) => {
  const { theme: contextTheme } = useContext(ThemeContext)
  return (
    <ThemeContextProvider theme={merge(celestialTheme, theme as Theme)}>
      <SCThemeProvider theme={contextTheme}>
        {children}
      </SCThemeProvider>
    </ThemeContextProvider>
  )
}

export default ThemeProvider
