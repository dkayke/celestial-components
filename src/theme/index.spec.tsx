import React from 'react'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import celestialTheme, {
  ThemeContextProvider,
  useThemeContext,
  withThemeProvider
} from '.'
import ThemeProvider from './ThemeProvider'

describe('Theme', () => {
  describe('ThemeProvider', () => {
    const ThemeMock = {
      color: { primary: { primary100: '#ff0000' } }
    }

    beforeEach(() => {
      jest.mock('react', () => ({
        ...jest.requireActual('react'),
        useContext: jest.fn(() => celestialTheme)
      }))
    })

    test('deve renderizar corretamente com um tema fornecido', () => {
      const { container } = render(
        <ThemeContextProvider theme={celestialTheme}>
          <ThemeProvider theme={ThemeMock}>
            <p>Celestial</p>
          </ThemeProvider>
        </ThemeContextProvider>
      )

      expect(container).toMatchSnapshot()
    })

    test('deve renderizar corretamente com o tema do contexto', () => {
      const { container } = render(
        <ThemeContextProvider theme={celestialTheme}>
          <ThemeProvider theme={celestialTheme}>
            <p>Celestial</p>
          </ThemeProvider>
        </ThemeContextProvider>
      )

      expect(container).toMatchSnapshot()
    })
  })

  describe('ThemeContext', () => {
    const MockedComponent: React.FC = () => {
      const context = useThemeContext()
      return <p aria-readonly={context?.loaded}>Celestial</p>
    }

    const WrappedComponent = withThemeProvider(MockedComponent)
    const WrappedComponentRendered = withThemeProvider(<MockedComponent /> as unknown as never)

    beforeEach(() => {
      jest.mock('react', () => ({
        ...jest.requireActual('react'),
        useContext: jest.fn(() => ({ theme: celestialTheme, loaded: true }))
      }))
    })

    test('renders without error when used inside ThemeContextProvider', () => {
      render(
        <ThemeContextProvider theme={celestialTheme}>
          <WrappedComponent />
        </ThemeContextProvider>
      )

      expect(screen.getByText('Celestial')).toBeInTheDocument()
    })

    test('throws an error when used outside CelestialProvider', () => {
      const originalError = console.error
      console.error = jest.fn()

      expect(() => render(<WrappedComponentRendered />)).toThrow(
        /Component must be used inside CelestialProvider/
      )

      expect(() => render(<WrappedComponent />)).toThrow(
        /MockedComponent must be used inside CelestialProvider/
      )

      console.error = originalError
    })

    test('throws an error when used outside ThemeContextProvider', () => {
      const originalError = console.error
      console.error = jest.fn()

      expect(() => render(<MockedComponent />)).toThrow(
        /useThemeContext must be used inside ThemeContextProvider/
      )

      console.error = originalError
    })
  })
})
