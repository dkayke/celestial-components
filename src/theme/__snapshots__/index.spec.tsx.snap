// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme ThemeProvider deve renderizar corretamente com o tema do contexto 1`] = `
<div>
  <p>
    Celestial
  </p>
</div>
`;

exports[`Theme ThemeProvider deve renderizar corretamente com um tema fornecido 1`] = `
<div>
  <p>
    Celestial
  </p>
</div>
`;
